src/Cli/Microsoft.DotNet.Cli.Utils/TypoCorrection.cs

```diff
@@ -7,40 +7,101 @@
+            var possibleSuggestions = possibleTokens.Select((string possibleMatch) => new Suggestion(possibleMatch, possibleMatch.Length));
+
+            var matchByStartsWith = possibleSuggestions
+               .Where(s => s.PossibleMatch.StartsWith(currentToken))
+               .OrderBy(s => s.Distance)
+               .ToList();
+
+            numberOfSuggestions += matchByStartsWith.Count;
+            if (numberOfSuggestions >= maxNumberOfSuggestions)
+            {
+                return matchByStartsWith
+                    .Take(maxNumberOfSuggestions)
+                    .Select(s => s.PossibleMatch);
+            }
+
+            var matchByContains = new List<Suggestion>();
+            if (currentToken.Length >= minCurrentTokenLength)
+            {
+                matchByContains = possibleSuggestions
+                    .Where(s => !matchByStartsWith.Contains(s)
+                        && s.PossibleMatch.Contains(currentToken)
+                        && s.Distance - currentTokenLength <= maxLevenshteinDistance)
+                    .OrderBy(s => s.Distance)
+                    .ToList();
+
+                numberOfSuggestions += matchByContains.Count;
+                if (numberOfSuggestions >= maxNumberOfSuggestions)
                 {
-                    (string _, int distance) = tuple;
-                    bestDistance ??= distance;
-                    return distance == bestDistance;
-                })
-                .Select(tuple => tuple.possibleMatch);
+                    return matchByStartsWith
+                        .Concat(matchByContains)
+                        .Take(maxNumberOfSuggestions)
+                        .Select(s => s.PossibleMatch);
+                }
+            }
+
+            var matchByLevenshteinDistance = possibleSuggestions
+                .Select(s => new Suggestion(s.PossibleMatch, GetDistance(s.PossibleMatch, currentToken)))
+                .Where(s => !(matchByStartsWith.Contains(s) || matchByContains.Contains(s)) && s.Distance <= maxLevenshteinDistance)
+                .OrderBy(s => s.Distance)
+                .ThenByDescending(s => GetStartsWithDistance(currentToken, s.PossibleMatch))
+                .FilterByShortestDistance()
+                .Take(maxNumberOfSuggestions - numberOfSuggestions);
+
+            return matchByStartsWith
+                .Concat(matchByContains
+                .Concat(matchByLevenshteinDistance))
+                .Select(s => s.PossibleMatch);
+        }
+
+        // The method takes the matches with the shortest distance
+        // e.g. (razor, 2), (pazor, 2), (pazors, 3) => (razor, 2), (pazor, 2)
+        private static IEnumerable<Suggestion> FilterByShortestDistance(this IEnumerable<Suggestion> possibleMatches)
+        {
+            int? bestDistance = null;
+
+            return possibleMatches.TakeWhile(s =>
+            {
+                (string _, int distance) = (s.PossibleMatch, s.Distance);
```

Comment:
nit: Why not simply:

```suggestion
                int distance = s.Distance;
```